Feature: Add file tree navigation to file detail page

samanehsan opened the pull request:
<h3>Purpose:</h3> 
Currently, there is no easy way to switch between files from the file detail page. Users must go back to the project dashboard or the file overview page and select the file they wish to view. This PR makes it easier to flip between files by adding the file tree as a toggle-able component on the file detail page. 

<h3>Changes:</h3> 
- Add a file navigation menu to the file detail page:
  - The file currently being viewed is highlighted in the grid** (if it is out of view, the grid scrolls to show it)
  - The panel is collapsable and resizes based on content  
  - Files are searchable 

![screen shot 2015-03-31 at 5 30 59 pm](https://cloud.githubusercontent.com/assets/6414394/6929817/e5c87f02-d7cb-11e4-8cf6-e17fd3535de5.png)

<h3>**Deferred issues:</h3>
- Highlighting selected files in the Box addon (requires building a path that includes folder/file names instead of ids) Related to #2356.
- Showing the dataverse file action button as either "released" or "draft" depending on the state of the file being viewed instead of always showing the "released" files as the default.
- Action buttons for github and dataverse get cut off with long repo/study names (#2369)

samanehsan commented:
This is ready for review @sloria :tada:

sloria commented on website/static/js/fileViewTreebeard.js:
```diff
@@ -0,0 +1,110 @@
+        } else {
+            tb.options.folderArray = [''];
+        }
+        m.render($('#files-search').get(0), tb.options.filterTemplate.call(tb));
```
Why is this call to `m.render` necessary? Shouldn't this be handled by Treebeard?

sloria replied:
> Use .lowercase-and-dashes for class names and #camelCase for IDs.

https://cosdev.readthedocs.org/en/latest/style_guides/html_css.html

samanehsan replied:
`m.render` is used here for rendering the search bar in the panel header, which is outside of the treebeard grid.

caneruguz replied:
The m.render use is a limitation of mithril, search is a mithril template that can be placed anywhere on the page but it needs to be added in mithril style so that the template can be interpreted. This is normally handled in Treebeard but in this case Saman want's to put the search bar in a different position. 

This use is not inefficient or wrong, it's how mithril templates are rendered into html while maintaining their bindings.

sloria commented on website/static/js/fileViewTreebeard.js:
```diff
@@ -0,0 +1,110 @@
+            tb.options.folderArray = [''];
+        }
+        m.render($('#files-search').get(0), tb.options.filterTemplate.call(tb));
+        $('#files-search input[placeholder=Search]').css('width', '95%');
```
Is it necessary to use jQuery for this? Can this go in CSS?

sloria replied:
Or, to be more specific: Could you override tb.filterTemplate and apply the styles on the returned mithril element rather than applying the styles with jQuery?

sloria commented on website/static/js/fileViewTreebeard.js:
```diff
@@ -0,0 +1,110 @@
+        m.render($('#files-search').get(0), tb.options.filterTemplate.call(tb));
+        $('#files-search input[placeholder=Search]').css('width', '95%');
+        $('#files-search input[placeholder=Search]').removeClass('pull-right').addClass('pull-left');
+        $('#toggle-icon').css('margin-top', '5px');
```
Same here; can this just be a CSS rule?

sloria commented on website/static/js/fileViewTreebeard.js:
```diff
@@ -0,0 +1,110 @@
+        }
+        m.render($('#files-search').get(0), tb.options.filterTemplate.call(tb));
+        $('#files-search input[placeholder=Search]').css('width', '95%');
+        $('#files-search input[placeholder=Search]').removeClass('pull-right').addClass('pull-left');
```
Is it possible to handle this in the mithril view for the search bar rather than doing manual CSS manipulation here?

samanehsan replied:
@sloria, do you mean adding this class in the mako file?

sloria commented on website/static/js/pages/view-file-page-dataverse.js:
```diff
@@ -0,0 +1,32 @@
+    $.ajax({
+        url: nodeApiUrl + 'files/grid/'
+    })
+        .done(function (data) {
```
Fix indentation.

sloria commented on website/static/js/pages/view-file-page-dataverse.js:
```diff
@@ -0,0 +1,32 @@
+});
+
+var panelToggle = $('.panel-toggle');
+    var panelExpand = $('.panel-expand');
```
Fix indentation.

sloria commented on website/static/js/pages/view-file-page-dataverse.js:
```diff
@@ -0,0 +1,32 @@
+    });
+});
+
+var panelToggle = $('.panel-toggle');
```
Can you add a `TODO` here noting that this should be there is duplication with the Wiki menu here and this should be refactored.

sloria commented on website/static/js/pages/view-file-page.js:
```diff
@@ -11,3 +14,33 @@ new FileRevisions(
+    $.ajax({
+        url: nodeApiUrl + 'files/grid/'
+    })
+        .done(function (data) {
```
Fix indentation.

sloria commented on website/static/js/pages/view-file-page.js:
```diff
@@ -11,3 +14,33 @@ new FileRevisions(
+            new fileBrowser(data);
+        });
+
+    var panelToggle = $('.panel-toggle');
```
This seems to be an exact copy of the dataverse view file page code. Can this be DRY'd up?

sloria commented on website/static/js/fangorn.js:
```diff
@@ -607,9 +607,12 @@ function _fangornLazyLoadOnLoad (tree) {
-    tree.sortChildren(this, sortDirection, 'text', 0);
-    this.redraw();
+    // Checking if this column does in fact have sorting
+    if (this.isSorted[0]) {
```
Why is `isSorted` and Array? @caneruguz

caneruguz replied:
It contains the sorted values for each of the possible columns that are sorted and its direction. It can be refactored so that the entire table has current sort state as the column like

```tb.isSorted : { col : 0, direction: 'asc' }```

This makes much more sense but let's talk about whether we want to implement it this way. It's not required for this PR to go forward.

sloria replied:
I found it unintuitive that `isSorted` was *not* a Boolean. I think this is a minor naming issue: perhaps `tb.sortOptions` would be more appropriate.

Not a big deal; this can be deferred.

samanehsan commented on website/static/js/fileViewTreebeard.js:
```diff
@@ -0,0 +1,110 @@
+    onload: function(tree) {
+        var tb = this;
+        Fangorn.DefaultOptions.onload.call(tb, tree);
+        $('.osf-panel-header.osf-panel-header-flex').show();
```
@sloria, would you mind clarifying what you mean? I'm not sure how to handle this in the view.

sloria replied:
We want to avoid manual DOM updates when possible and prefer to declare in the view. Something along the lines of:

```js
m('.osf-panel-header.osf-panel-header-flex', {
    style: {display: ctrl.isVisible ? 'block' : 'none'}
});
```

Where `ctrl` is the controller for the panel.

I haven't given much thought to implementation in this particular case, but I'm wondering if there's opportunity to clean this up.

sloria commented on website/templates/project/addon/view_file.mako:
```diff
@@ -41,11 +69,30 @@
+                }
+        });
+        </script>
+        <script src=${"/static/public/js/view-file-tree-page.js" | webpack_asset}></script>
```
Why is this in the `else` block? Shouldn't this script be loaded even if rendered is not `None`?

sloria commented on website/templates/project/view_file.mako:
```diff
@@ -12,7 +17,34 @@
+                    <div class="fangorn-loading"> <i class="fa fa-spinner fangorn-spin"></i> <p class="m-t-sm fg-load-message"> Loading files...  </p> </div>
+                </div>
+            </div>
+        </div>
```
Fix indentation.

sloria commented on website/templates/project/view_file.mako:
```diff
@@ -12,7 +17,34 @@
+        </div>
+
+    <!-- Menu toggle closed -->
+            <div class="osf-panel panel-collapsed hidden-xs text-center reset-height"  style="display: none">
```
Fix indentation.

sloria commented on website/static/js/fileViewTreebeard.js:
```diff
@@ -0,0 +1,115 @@
+    }
+
+    var fangornOpts = {
+    divID: 'grid',
```
Fix indentation.

sloria commented on website/static/js/pages/view-file-tree-page.js:
```diff
@@ -0,0 +1,34 @@
+var fileBrowser = require('../fileViewTreebeard.js');
```
Omit the `.js` extension.

sloria replied:
Missing `use strict;`.

sloria commented:
@samanehsan was looking into a known bug with this PR where the file tree would scroll to its initial position after expanding a folder. To reproduce:

- On a project with lots of files and least one folder within an addon, view a file.
- On the view file page, the file tree should be scrolled to the position of the currently-viewed file.
- In the file tree, scroll to a folder and expand it

The file tree unexpectedly scrolls back to the currently-viewed file.

@caneruguz Can you look into this?

sloria commented on website/static/js/fangorn.js:
```diff
@@ -833,6 +836,45 @@ function expandStateLoad(item) {
+ * @param nodeID Current node._id
+ * @param file window.contextVars.file object
+ */
+function findCurrentFileID(tree, nodeID, file) {
```
This name is a bit of a misnomer:  `find*` implies that this is doing some sort of search and returning an ID, but it is not. It has a number of side effects, updating `folderIndex`, updating folders, etc.

samanehsan replied:
@sloria would `getCurrentFileID` be a more appropriate name? I'm not updating `folderIndex`, just retrieving it so that the scroll location can be calculated. The update folders is for the purpose of getting the file id (if it is in a subfolder whose data has not been loaded yet).

sloria replied:
Yes, I think `setCurrentFileID` makes sense.

`folderIndex` appears to be updated on line 850.

samanehsan replied:
Oh, you're right! I got this confused with something else.

sloria commented on website/static/js/fangorn.js:
```diff
@@ -833,6 +836,45 @@ function expandStateLoad(item) {
+ */
+function findCurrentFileID(tree, nodeID, file) {
+    var tb = this;
+    if (tb.options.folderIndex !== undefined && tb.options.folderArray !== undefined && tb.options.folderIndex < tb.options.folderArray.length) {
```
@samanehsan Do these "folder" variables need to be on `tb.options`. In general, `tb.options` should never be changed. Also, those `options` are configuration options that alter the behavior of treebeard; including other variables is confusing.

samanehsan replied:
I thought in that in general we did not want to put too many variables directly onto `tb` but I may have misunderstood. What do you think @caneruguz?

sloria replied:
You can add a namespace to avoid pollution.

```js
tb._fangorn = {};
tb._fangorn.folderIndex = 123;
```

caneruguz replied:
I don't know, this discussion can become academic; what is important is that the change you make is available to you within other aspects of treebeard. TB itself will always be available and hence tb.options as well but tb.options is the only way to inject anything into Treebeard. So you need a place to add your namespaced variables at such a time that they will be available. For instance if you add it in onload option you will only have it applied to Tb when it gets loaded. I hope this was clear, so the point is you don't have access to tb when you are setting options but your options hooks (like onload or others that are methods) have it.

samanehsan replied:
@sloria @caneruguz, yeah that seems to be the problem I'm running into. I tried setting `tb._fangorn` in `ondataload` but it is not available to `lazyLoadOnLoad`

sloria replied:
Would be nice if there was an initialization hook:

```js
var tb = new Treebeard({
    init: function() {
        this._fangorn = {};
    }
});
```

@caneruguz Is something like this available?

caneruguz replied:
Hmm... I see, yes that could be added with the revision of TB constructor.

sloria replied:
For the time being, we can just use a variable name prefix:

```js
tb.filepickerFolderIndex = ...
# or
tb.fangornFolderIndex = ...
```

caneruguz commented:
Potential solution to this comment: https://github.com/CenterForOpenScience/osf.io/pull/2229#issuecomment-91356076

is here: https://github.com/samanehsan/osf.io/pull/6

Let me know what you guys think.

sloria commented on website/static/js/pages/view-file-page.js:
```diff
@@ -1,3 +1,5 @@
+var nodeApiUrl = window.contextVars.node.urls.api;
```
This variable is unused.

sloria commented:
Once @caneruguz 's PR is updated and merged, this will be ready for final pass.

sloria commented on website/static/js/fangorn.js:
```diff
@@ -833,6 +836,45 @@ function expandStateLoad(item) {
+        var visibleIndex = tb.visibleIndexes.indexOf(index);
+        if (visibleIndex !== -1 && visibleIndex > tb.showRange.length - 2) {
+            var scrollTo = visibleIndex * tb.options.rowHeight;
+            $('#tb-tbody').scrollTop(scrollTo);
```
@caneruguz points out that this is incorrect in https://github.com/caneruguz/treebeard/issues/76#issuecomment-93051613 . 

@caneruguz Could you fix this in your PR?

sloria replied:
This still has not been changed. Is this supposed to be `this.select('#tb-tbody').scrollTop(scrollTo)`, @caneruguz ?

samanehsan commented:
@sloria, I added in the code comment about events and changed the treebeard selector to use `this.select()` Do you want me to pull from develop as well?

sloria commented:
@samanehsan Thanks. I'll go ahead and create a new branch for this that is up to date with develop.

samanehsan commented:
Thanks @sloria! :airplane:

samanehsan commented:
Ahhh accidentally hit "close & comment" instead of "comment"

caneruguz commented:
Saman can you please change that segment, you are right it should be
tb.select(.....

On Wed, Apr 15, 2015 at 1:06 PM, Saman Ehsan <notifications@github.com>
wrote:

> Ahhh accidentally hit "close & comment" instead of "comment"
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/CenterForOpenScience/osf.io/pull/2229#issuecomment-93488244>
> .
>

samanehsan commented:
@caneruguz, I changed the `#tb-tbody` selector to `this.select('#tb-tbody')`

sloria commented:
Merged this feature in https://github.com/CenterForOpenScience/osf.io/commit/5f61d5445c7b95516ed3a8937cf435793ad8bd00
